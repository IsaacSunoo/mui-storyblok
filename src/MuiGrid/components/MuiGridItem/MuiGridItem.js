import React, { createElement, lazy, Suspense } from 'react';
import PropTypes from 'prop-types';
import Grid from '@material-ui/core/Grid';
import Storyblok from '../../../utils/Storyblok';

const MuiButtonRedirect = lazy(() => import('../../../MuiButtonRedirect/MuiButtonRedirect'));
const MuiCard = lazy(() => import('../../../MuiCard/MuiCard'));
const MuiTypography = lazy(() => import('../../../MuiTypography/MuiTypography'));
const MuiMenu = lazy(() => import('../../../MuiMenu/MuiMenu'));
const MuiIcon = lazy(() => import('../../../MuiIcon/MuiIcon'));
const MuiExpansionPanel = lazy(() => import('../../../MuiExpansionPanel/MuiExpansionPanel'));
// const MuiList = lazy(() => import('../../../MuiList/MuiList'));

const MuiGridItem = ({
  sizeGrid,
  content,
  direction,
  justify,
  alignItems,
  xl,
  lg,
  md,
  sm,
  xs,
  spacing,
  rootClass,
}) => {
  const components = {
    MuiButtonRedirect,
    MuiCard,
    MuiTypography,
    MuiMenu,
    MuiIcon,
    MuiExpansionPanel,
    // MuiList,
  };

  const styles = Storyblok.arrayToMuiStyles(rootClass, { padding: '25px' });

  return (
    <Grid
      className={styles.root}
      container
      item
      direction={direction}
      justify={justify}
      alignItems={alignItems}
      spacing={(Number(spacing))}
      xl={sizeGrid(xl)}
      lg={sizeGrid(lg)}
      md={sizeGrid(md)}
      sm={sizeGrid(sm)}
      xs={sizeGrid(xs)}
    >
      <Suspense fallback={<div />}>
        {content.map((item, index) => createElement(
          components[item.component],
          Object.assign(item, { key: index }),
        ))}
      </Suspense>
    </Grid>
  );
};

export default MuiGridItem;

MuiGridItem.propTypes = {
  // mui prop: 'row' | 'row-reverse' | 'column' | 'column-reverse'
  direction: PropTypes.string,
  // mui prop: 'flex-start' | 'center' | 'flex-end' | 'space-between' | 'space-around' | 'space-evenly'
  justify: PropTypes.string,
  // mui prop: 'flex-start' | 'center' | 'flex-end' | 'stretch' | 'baseline'
  alignItems: PropTypes.string,
  // mui prop:  0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10
  spacing: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  // mui prop: false | 'auto' | true | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12
  xl: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.bool]),
  // mui prop: false | 'auto' | true | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12
  lg: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.bool]),
  // mui prop: false | 'auto' | true | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12
  md: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.bool]),
  // mui prop: false | 'auto' | true | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12
  sm: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.bool]),
  // mui prop: false | 'auto' | true | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12
  xs: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.bool]),

  /** prop from parent compoent to render corect type for grid size */
  sizeGrid: PropTypes.func.isRequired,

  /** stroyblok multiselect of css classes */
  rootClass: PropTypes.arrayOf(PropTypes.string),

  content: PropTypes.arrayOf(PropTypes.shape({
    component: PropTypes.string.isRequired,
  })).isRequired,
};

MuiGridItem.defaultProps = {
  rootClass: [],
  xl: false,
  lg: false,
  md: false,
  sm: false,
  xs: false,
  direction: 'row',
  justify: 'center',
  alignItems: 'center',
  spacing: 2,
};
