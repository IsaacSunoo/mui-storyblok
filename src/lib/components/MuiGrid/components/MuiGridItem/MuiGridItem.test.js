import React from 'react';
import { shallow } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import MuiGridItem from './MuiGridItem';

function setup() {
  const props = {
    sizeGrid: jest.fn(),
    content: [{
      component: 'MuiIcon',
      iconName: 'android',
    }, {
      component: 'MuiButtonRedirect',
      redirectRoute: 'page-welcome',
      button: [{
        component: 'MuiButton',
        buttonText: 'buttonText',
      }],
    }, {
      component: 'MuiCard',
      content: [{
        component: 'MuiCardMedia',
        alt: 'alt',
        title: 'title',
        image: 'img.com',
      }],
    }, {
      component: 'MuiTypography',
      content: [{
        component: 'MuiText',
        text: 'text',
      }],
    }, {
      component: 'MuiMenu',
      btnText: 'btnText',
      content: [{
        component: 'MuiMenuItem',
        buttonText: 'buttonText',
        redirectRoute: 'page-welcome',
      }],
    }, {
      component: 'MuiExpansionPanel',
      expansionPanelSummary: [{
        component: 'MuiExpansionPanelSummary',
        content: [{
          component: 'MuiIcon',
          iconName: 'android',
        }],
        expandIcon: [{
          component: 'MuiIcon',
          iconName: 'android',
        }],
      }],
      expansionPanelDetails: [{
        component: 'MuiExpansionPanelDetails',
        content: [{
          component: 'MuiTypography',
          content: [{
            component: 'MuiText',
            text: 'text',
          }],
        }],
      }],
    }],
  };
  const comp = shallow(<MuiGridItem {...props} />);
  return { comp, props };
}

describe('<MuiGridItem />', () => {
  it('renders MuiGridItem', () => {
    const { comp } = setup();
    expect(comp).toBeDefined();
  });

  test('snapshot', () => {
    const { props } = setup();
    const tree = renderer.create((
      <MemoryRouter>
        <MuiGridItem {...props} />
      </MemoryRouter>
    ));
    expect(tree).toMatchSnapshot();
  });
});
