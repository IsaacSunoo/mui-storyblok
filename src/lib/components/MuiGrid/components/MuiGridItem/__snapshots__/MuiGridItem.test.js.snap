// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MuiGridItem /> snapshot 1`] = `
<div
  className="MuiGrid-root makeStyles-root-3 MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-item MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
>
  <span
    aria-hidden={true}
    className="material-icons MuiIcon-root makeStyles-root-107 MuiIcon-colorSecondary"
  >
    android
  </span>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-root-117 MuiButton-textSecondary MuiButton-fullWidth"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-label"
    >
      buttonText
    </span>
  </button>
  <div
    className="MuiGrid-root makeStyles-root-150"
    raised="false"
  >
    <div
      alt="alt"
      className="MuiCardMedia-root makeStyles-root-151"
      style={
        Object {
          "backgroundImage": "url(\\"img.com\\")",
        }
      }
      title="title"
    />
  </div>
  <p
    className="MuiTypography-root makeStyles-root-155 MuiTypography-body1"
  >
    text
  </p>
  <div />
  <div
    className="MuiPaper-root MuiExpansionPanel-root makeStyles-root-187 MuiExpansionPanel-rounded MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      aria-disabled={false}
      aria-expanded={false}
      className="MuiButtonBase-root MuiExpansionPanelSummary-root makeStyles-root-220"
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="MuiExpansionPanelSummary-content"
      >
        <span
          aria-hidden={true}
          className="material-icons MuiIcon-root makeStyles-root-227 MuiIcon-colorSecondary"
        >
          android
        </span>
      </div>
      <div
        aria-disabled={false}
        aria-hidden={true}
        className="MuiButtonBase-root MuiIconButton-root MuiExpansionPanelSummary-expandIcon MuiIconButton-edgeEnd"
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role={null}
        tabIndex={null}
      >
        <span
          className="MuiIconButton-label"
        >
          <span
            aria-hidden={true}
            className="material-icons MuiIcon-root makeStyles-root-237 MuiIcon-colorSecondary"
          >
            android
          </span>
        </span>
      </div>
    </div>
    <div
      className="MuiCollapse-container MuiCollapse-hidden"
      style={
        Object {
          "minHeight": "0px",
        }
      }
    >
      <div
        className="MuiCollapse-wrapper"
      >
        <div
          className="MuiCollapse-wrapperInner"
        >
          <div
            role="region"
          >
            <div
              className="MuiExpansionPanelDetails-root makeStyles-root-243"
            >
              <p
                className="MuiTypography-root makeStyles-root-245 MuiTypography-body1"
              >
                text
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
